rust_snuba/src/processors/replays.rs

```diff
@@ -11,145 +11,166 @@ pub fn process_message(
+            })
+            .collect()),
         ReplayPayload::Event(event) => {
             let event_hash = match (event.event_hash, event.segment_id) {
                 (None, None) => Uuid::new_v4(),
                 (None, Some(segment_id)) => {
-                    Uuid::from_slice(md5::compute(segment_id.to_string().as_bytes()).as_slice())?
+                    Uuid::from_slice(md5::compute(segment_id.to_string().as_bytes()).as_slice())
+                        .expect("md5 hash never exceeds 16 bytes")
                 }
                 (Some(event_hash), _) => event_hash,
             };
 
-            let (ip_address_v4, ip_address_v6) = match event.user.ip_address {
-                None => (None, None),
-                Some(IpAddr::V4(ip)) => (Some(ip), None),
-                Some(IpAddr::V6(ip)) => (None, Some(ip)),
-            };
-
             // Tags normalization and title extraction.
             let mut title = None;
             let mut tags_key = Vec::with_capacity(event.tags.len());
             let mut tags_value = Vec::with_capacity(event.tags.len());
 
             for tag in event.tags.into_iter() {
                 if &tag.0 == "transaction" {
-                    title = Some(tag.1.clone())
+                    title = tag.1.clone()
                 }
                 tags_key.push(tag.0);
                 tags_value.push(tag.1);
             }
 
-            // Compute user value.
-            let user = if !event.user.user_id.is_empty() {
-                event.user.user_id.clone()
-            } else if !event.user.username.is_empty() {
-                event.user.username.clone()
-            } else if !event.user.email.is_empty() {
-                event.user.email.clone()
-            } else if let Some(ip) = event.user.ip_address {
-                ip.to_string()
-            } else {
-                String::new()
+            // Unwrap the ip-address string.
+            let ip_address_string = event.user.ip_address.unwrap_or_default();
+            let (ip_address_v4, ip_address_v6) = match ip_address_string.parse::<IpAddr>() {
+                Err(_) => (None, None),
+                Ok(ip) => match ip {
+                    IpAddr::V4(ipv4) => (Some(ipv4), None),
+                    IpAddr::V6(ipv6) => (None, Some(ipv6)),
+                },
+            };
+
+            // Derive user value through precedence checks.
+            let user = match (
+                &event.user.user_id,
+                &event.user.username,
+                &event.user.email,
+                &ip_address_v4,
+                &ip_address_v6,
+            ) {
+                (Some(uid), _, _, _, _) => uid.clone(),
+                (_, Some(username), _, _, _) => username.clone(),
+                (_, _, Some(email), _, _) => email.clone(),
+                (_, _, _, Some(ip), _) => ip.to_string().clone(),
+                (_, _, _, _, Some(ip)) => ip.to_string().clone(),
+                _ => String::new(),
```

Comment:
I am very unhappy with this code.  Its a precedence check.  In python it looks like this `user = user_id or user_email or ""`.  Where the bool(user_id) evaluates to false if `None` or false if `""`.  The first non-empty string is accepted.  Any help cleaning this up is appreciated.

Reply:
You can chain `Option::or` without a problem, like so:

```rust
event.user.user_id.as_ref().or(event.user.username.as_ref())
.or(event.user.email.as_ref()).unwrap_or_default()
```

The compiler will probably complain about types and references, so worst case you would have to clone everything, which can also be cleaned up at a later point.